### PR TITLE
Coerce np.matrix instances to np.ndarray

### LINKED ISSUE #1431: Seeming incompatibility with the numpy matrix subclass
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of scanpy.
- [ ] (optional) I have confirmed this bug exists on the master branch of scanpy.

### What happened?

For reasons I could not explain, I have found that the method `filter_genes` causes a `ValueError` when the `adata.X` object is of the type `numpy.matrix`. It is easy to circumvent by converting it to a general `ndarray`, but I wanted to file it here for reference, as `matrix` objects are still given by default by some methods (such as the `todense()` method of a sparse matrix) and `matrix` is a subclass of `ndarray` so it is not easy to identify it as a type error. Here is a minimal code sample

### Minimal code sample

```python
import scanpy as sc
import anndata
import numpy as np
import pandas as pd
X = np.matrix([[1, 2], [3, 0]])
print(isinstance(X, np.ndarray))
ad = anndata.AnnData(X=X, obs={'obs_names': ['a', 'b']}, var={'vars_names': ['v1', 'v2']})
sc.pp.filter_genes(ad, min_cells=2)
```

### Error output

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[47], line 9
      7 print(isinstance(X, np.ndarray))
      8 ad = anndata.AnnData(X=X, obs={'obs_names': ['a', 'b']}, var={'vars_names': ['v1', 'v2']})
----> 9 sc.pp.filter_genes(ad, min_cells=2)

File ~/miniconda3/envs/bio/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:250, in filter_genes(data, min_counts, min_cells, max_counts, max_cells, inplace, copy)
    248     adata.var['n_counts'] = number
    249 else:
--> 250     adata.var['n_cells'] = number
    251 adata._inplace_subset_var(gene_subset)
    252 return adata if copy else None

File ~/miniconda3/envs/bio/lib/python3.10/site-packages/pandas/core/frame.py:3950, in DataFrame.__setitem__(self, key, value)
   3947     self._setitem_array([key], value)
   3948 else:
   3949     # set column
-> 3950     self._set_item(key, value)

File ~/miniconda3/envs/bio/lib/python3.10/site-packages/pandas/core/frame.py:4143, in DataFrame._set_item(self, key, value)
   4133 def _set_item(self, key, value) -> None:
   4134     """
   4135     Add series to DataFrame in specified column.
   4136 
   (...)
   4141     ensure homogeneity.
   4142     """
-> 4143     value = self._sanitize_column(value)
   4145     if (
   4146         key in self.columns
   4147         and value.ndim == 1
   4148         and not is_extension_array_dtype(value)
   4149     ):
   4150         # broadcast across multiple columns if necessary
   4151         if not self.columns.is_unique or isinstance(self.columns, MultiIndex):

File ~/miniconda3/envs/bio/lib/python3.10/site-packages/pandas/core/frame.py:4870, in DataFrame._sanitize_column(self, value)
   4867     return _reindex_for_setitem(Series(value), self.index)
   4869 if is_list_like(value):
-> 4870     com.require_length_match(value, self.index)
   4871 return sanitize_array(value, self.index, copy=True, allow_2d=True)

File ~/miniconda3/envs/bio/lib/python3.10/site-packages/pandas/core/common.py:576, in require_length_match(data, index)
    572 """
    573 Check the length of data matches the length of the index.
    574 """
    575 if len(data) != len(index):
--> 576     raise ValueError(
    577         "Length of values "
    578         f"({len(data)}) "
    579         "does not match length of index "
    580         f"({len(index)})"
    581     )

ValueError: Length of values (1) does not match length of index (2)
```

### Versions

<details>

```
-----
anndata     0.10.2
scanpy      1.9.6
-----
PIL                 9.4.0
appnope             0.1.3
asttokens           NA
backcall            0.2.0
bottleneck          1.3.5
cffi                1.15.1
comm                0.1.3
cycler              0.10.0
cython_runtime      NA
dateutil            2.8.2
debugpy             1.6.7
decorator           5.1.1
exceptiongroup      1.1.3
executing           1.2.0
google              NA
h5py                3.9.0
ipykernel           6.23.1
jedi                0.18.2
joblib              1.3.2
kiwisolver          1.4.4
llvmlite            0.41.1
matplotlib          3.7.2
mkl                 2.4.0
mpl_toolkits        NA
natsort             8.4.0
numba               0.58.1
numexpr             2.8.4
numpy               1.25.2
packaging           23.1
pandas              2.0.3
parso               0.8.3
pexpect             4.8.0
pickleshare         0.7.5
pkg_resources       NA
platformdirs        3.5.1
prompt_toolkit      3.0.38
psutil              5.9.0
ptyprocess          0.7.0
pure_eval           0.2.2
pydev_ipython       NA
pydevconsole        NA
pydevd              2.9.5
pydevd_file_utils   NA
pydevd_plugins      NA
pydevd_tracing      NA
pygments            2.15.1
pyparsing           3.0.9
pytz                2022.7
scipy               1.11.1
session_info        1.0.0
six                 1.16.0
sklearn             1.3.0
stack_data          0.6.2
threadpoolctl       3.2.0
tornado             6.3.2
traitlets           5.9.0
typing_extensions   NA
wcwidth             0.2.6
yaml                6.0.1
zmq                 25.1.0
zoneinfo            NA
-----
IPython             8.14.0
jupyter_client      8.2.0
jupyter_core        5.3.0
-----
Python 3.10.10 (main, Mar 21 2023, 13:41:39) [Clang 14.0.6 ]
macOS-10.16-x86_64-i386-64bit
-----
Session information updated at 2024-01-24 17:43

```

</details>

### PR DESCRIPTION
**Pull request by flying-sheep**
- [x] Closes #1431
- [x] Tests added
- [x] Release note added (or unnecessary)

This unifies the array type coercion, which used to work differently betwen `X` and `AlignedMapping`s.

@ilan-gold, does it make sense to add AwkwardArray there or should we special-case it so it’s only valid in {obs,var}m, not in X and layers?

### CODE REVIEW

File: src/anndata/_core/anndata.py [hunk 1]
```diff
@@ -419,12 +390,12 @@ def _init_as_actual(
...
-                if isinstance(X, s_type.value):
-                    break
-            else:
+            if not isinstance(X, StorageType.classes()):
                 raise ValueError(
-                    f"X needs to be of one of numpy.ndarray, numpy.ma.core.MaskedArray, scipy.sparse.spmatrix, h5py.Dataset, zarr.Array, anndata.experimental.[CSC,CSR]Dataset, dask.array.Array, cupy.ndarray, cupyx.scipy.sparse.spmatrix, not {type(X)}."
+                    "X needs to be of one of numpy.ndarray, numpy.ma.core.MaskedArray, "
+                    "scipy.sparse.spmatrix, h5py.Dataset, zarr.Array, "
+                    "anndata.experimental.[CSC,CSR]Dataset, dask.array.Array, "
+                    f"cupy.ndarray, cupyx.scipy.sparse.spmatrix, not {type(X)}."
```
Comment (flying-sheep) [inline src/anndata/_core/anndata.py:398]: maybe this check should be moved into `coerce_arrays`?

also TODO: construct error message from enum

File: src/anndata/_core/storage.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        return value if allow_df else ensure_df_homogeneous(value, name)
+    # if value is an array-like object, try to convert it
+    if allow_array_like:
+        try:
+            # TODO: asarray? asanyarray?
+            return np.array(value)
+        except (ValueError, TypeError):
+            pass
+    # if value isn’t the right type or convertible, raise an error
+    msg = f"X needs to be of one of {join_english(StorageType.qualnames())}, not {type(value)}."
```
Comment (ilan-gold) [inline src/anndata/_core/storage.py:84]: Should the message include the possibility that it is not convertible?

File: tests/test_base.py [hunk 3]
```diff
@@ -168,6 +166,39 @@ def test_df_warnings():
...
+            setattr(adata, attr, mat)
+        else:
+            getattr(adata, attr)["a"] = mat
+    else:
+        raise ValueError(when)
+
+    # with pytest.warns(ImplicitModificationWarning, match=r"np\.ndarray"):
+    arr = getattr(adata, attr) if direct else getattr(adata, attr)["a"]
+    assert isinstance(arr, np.ndarray), "it’s not even an array"
+    assert not isinstance(arr, np.matrix), "it’s still a matrix"
```
Comment (ilan-gold) [inline tests/test_base.py:199]:
```suggestion
    assert isinstance(arr, np.ndarray), f"{arr} is not an array"
    assert not isinstance(arr, np.matrix), f"{arr} is still a matrix"
```

File: src/anndata/_core/storage.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    # If value is one of the allowed types, return it
+    if isinstance(value, StorageType.classes()):
+        return value.A if isinstance(value, np.matrix) else value
+    if isinstance(value, pd.DataFrame):
+        return value if allow_df else ensure_df_homogeneous(value, name)
+    # if value is an array-like object, try to convert it
+    if allow_array_like:
+        try:
+            # TODO: asarray? asanyarray?
+            return np.array(value)
```
Comment (ilan-gold) [inline src/anndata/_core/storage.py:80]: Should we raise https://anndata.readthedocs.io/en/latest/generated/anndata.ImplicitModificationWarning.html?  I see it commented out below

File: src/anndata/_core/storage.py (hunk 2 above)
Comment (flying-sheep) [reply src/anndata/_core/storage.py:84]: Good idea I’ll use exception hierarchy for this (`rais A() from b`)

File: src/anndata/_core/storage.py (hunk 4 above)
Comment (flying-sheep) [reply src/anndata/_core/storage.py:80]: I added it explicitly for matrices, but if you do `AnnData([[1,2], [3,4]])` you shouldn’t get a warning

Comment (flying-sheep) [reply src/anndata/_core/storage.py:80]: We’re using this code also when doing e.g. `adata[:2, 0].X = [0, 0]`, in which case we don’t want it, so I added it only for the matrix case, not around the line where you commented.

**Review by ilan-gold — APPROVED** [review]

**Pull request merged**